Enable nullable references in Moq.csproj

andrewimcclement opened the pull request:
Aims to resolve #1418 , #1416 .

Does not extend to Moq.Tests.csproj, etc.

Nullable reference issues are enabled as warnings in Moq.csproj. Several issues have been left unresolved, as I lack the knowledge to resolve them correctly at this time (perhaps maintainers can help here, or else it may be acceptable to put this PR down as is and slowly pick off the remaining nullable reference warnings one at a time).

In general I have tried hard to avoid actually changing the code functionality, except where it seemed clear. So several simplifications could be made in follow up PRs.

See [the docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis) for details of the nullability attributes I have used.

andrewimcclement committed:
Enable nullable warnings in Moq.csproj.

andrewimcclement committed:
Fix up nullability of ExpressionComparer.cs.

andrewimcclement committed:
Fix up nullability in ActionObserver.cs.

andrewimcclement committed:
Fix up nullability in Match.cs.

andrewimcclement committed:
Fix up nullability in IProtectedAsMock.cs.

andrewimcclement committed:
Fix up nullability in MethodExpectation.cs.

andrewimcclement committed:
Fix up some nullable issues in ProtectedAsMock.cs.

andrewimcclement committed:
Fix up nullability in AwaitableFactory*.cs.

andrewimcclement committed:
Fix up nullability in Awaitable.cs.

andrewimcclement committed:
Fix up nullability in TypeMatcherAttribute.cs.

andrewimcclement committed:
Fix up nullability in Extensions.cs. A new symbol has been defined for ease of use of nullability checks.

andrewimcclement committed:
Fix up nullability in RaiseEvent.cs.

andrewimcclement committed:
Fix up nullability in Expectation.cs and MethodExpectation.cs.

andrewimcclement committed:
Fix up AwaitableFactory`2.CreateCompleted(object?) to correctly assume non-nullability.

andrewimcclement committed:
Simplify TaskFactory.CreateCompleted.

andrewimcclement committed:
Fix nullability for AsInterface.cs.

andrewimcclement committed:
Fix up nullability in EventHandlerCollection.cs.

andrewimcclement committed:
Fix up nullable types in ExpressionReconstructor.cs.

andrewimcclement committed:
Fix nullability in Condition.cs. Add guard clause in Mock.When as nothing ever calls `new Condition(null)` so we can simply rely on nullable annotations, so long as we guard the public interface.

andrewimcclement committed:
Fix up nullability in StubbedPropertySetup.cs.

andrewimcclement committed:
Fix up nullability in StubbedPropertiesSetup.cs.

andrewimcclement committed:
Fix up nullability in Invocation.cs.

andrewimcclement committed:
Fix up nullability in Extensions.cs.

andrewimcclement committed:
Fix nullability in InnerMockSetup.cs.

andrewimcclement committed:
Fix up nullability in StubbedPropertySetup.cs.

andrewimcclement committed:
Fix nullability in Setup.cs.

andrewimcclement committed:
Fix nullability in SetupCollection.cs.

andrewimcclement committed:
Fix up nullability in StringBuilderExtensions.cs.

andrewimcclement committed:
Fix up nullability in AwaitableFactory.cs.

andrewimcclement committed:
Fix up nullability in Times.cs.

andrewimcclement committed:
Fix up nullability in EmptyDefaultValueProvider.cs.

andrewimcclement committed:
Fix up TaskFactory`1.cs nullability.

andrewimcclement committed:
Fix up nullability in ActionObserver.cs.

andrewimcclement committed:
Fix up nullability in ValueTaskFactory`1.cs.

andrewimcclement committed:
Fix up nullability for Invocation.cs.

andrewimcclement committed:
Tidy nullability in UpgradePropertyAccessorMethods.cs.

andrewimcclement committed:
Fix up nullability in Extensions.cs.

andrewimcclement committed:
Tidy nullability in AwaitableFactory`2.cs.

andrewimcclement committed:
Fix up nullability in DefaultValueProvider.cs.

andrewimcclement committed:
Fix up nullability in LookupOrFallbackDefaultValueProvider.cs.

andrewimcclement committed:
Fix up nullability in Match.cs.

andrewimcclement committed:
Tidy nullability in ReturnBaseOrDefaultValue.cs.

andrewimcclement committed:
Fix up nullability in InvocationCollection.cs.

andrewimcclement committed:
Tidy nullability in Guard.cs.

andrewimcclement committed:
Fix up nullability in RaiseEvent.cs.

andrewimcclement committed:
Fix up nullability in EmptyDefaultValueProvider.cs.

andrewimcclement committed:
Fix up nullability in Pair.cs

andrewimcclement committed:
Fix up nullability in SetupWithOutParameterSupport.cs.

andrewimcclement committed:
Fix up nullability in CastleProxyFactory.cs.

andrewimcclement committed:
Fix up a couple more Setup classes.

andrewimcclement committed:
Tidy nullability in Match.cs.

andrewimcclement committed:
Tidy nullability in InterfaceProxy.cs.

andrewimcclement committed:
Fix up nullability in ReturnValue.cs.

andrewimcclement committed:
FIx up nullability in EvaluateCaptures.cs.

andrewimcclement committed:
Tidy nullability in RefMatcher.cs.

andrewimcclement committed:
Fix up nullability in IMatcher.cs and derived types.

andrewimcclement committed:
Tidy nullability in Match.cs.

andrewimcclement committed:
Tidy nullability in MatcherFactory.cs.

andrewimcclement committed:
Tidy nullability in MatcherObserver.cs.

andrewimcclement committed:
Fix up nullability in MockException.cs.

andrewimcclement committed:
Fix up nullability in ParamArrayMatcher.cs.

andrewimcclement committed:
Fix up nullability in ReturnComputedValue.cs.

andrewimcclement committed:
Tidy nullability in SetupPhrase.cs.

andrewimcclement committed:
Make best guess at how ThrowComputedException.cs should work.

andrewimcclement committed:
Tidy nullability in Mock.cs.

andrewimcclement committed:
Tidy nullability in ActionObserver.cs.

andrewimcclement committed:
Tidy nullability in StubbedPropertiesSetup.cs.

andrewimcclement committed:
Tidy nulability in Capture.cs.

andrewimcclement committed:
Tidy nullability in ProtectedAsMock.cs and ProtectedMock.cs.

andrewimcclement committed:
Tidy nullability in MockDefaultValueProvider.cs.

andrewimcclement committed:
Improve nullability in MethodCall.cs - still some strangeness left over.

andrewimcclement committed:
Tidy nullability in Guard.cs.

andrewimcclement committed:
Tidy nullability in CastleProxyFactory.cs, InterceptionAspects.cs and ProxyFactory.cs.

andrewimcclement committed:
Fix up nullability in MatcherAttributeMatcher.cs.

andrewimcclement committed:
Tidy nullability in ActionObserver.cs.

andrewimcclement committed:
Tidy nullability in MockSetupsBuilder.cs.

andrewimcclement committed:
Tidy nullability in StringBuilderExtensions.AppendExpression.cs.

andrewimcclement committed:
Tidy nullability in ExpressionReconstructor.cs.

andrewimcclement committed:
Tidy nullability of Evaluator.cs.

andrewimcclement committed:
Tidy nullability in ItExpr.cs.

andrewimcclement committed:
Tidy nullability in MockFactory.cs.

andrewimcclement committed:
Tidy AsInterface.cs.

andrewimcclement committed:
Tidy ConstructorCallVisitor.cs. Fields `constructor` and `arguments` still need dealing with.

andrewimcclement committed:
Tidy ReturnsExtensions.cs nullability.

andrewimcclement committed:
Tidy nullability in Mock`1.cs.

andrewimcclement committed:
Tidy nullability in MockException.cs.

andrewimcclement committed:
Partially fix up nullability in ProtectedMock.cs.

andrewimcclement committed:
Tidy nullability in Mock.cs.

andrewimcclement committed:
Some nullability improvements to ExpressionExtensions.cs. It seems like nullability is a little dodgy here.

andrewimcclement committed:
Tidy nullability in It.cs.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/devlooped/moq?pullRequest=1535) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/devlooped/moq?pullRequest=1535) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/devlooped/moq?pullRequest=1535) it.</sub>

andrewimcclement committed:
Revert change to Recorder.invocationTimestamp.

andrewimcclement committed:
Dotnet format.

andrewimcclement commented:
Run the dotnet format command.

andrewimcclement committed:
Simplify StringBuilder.AppendValueOf handling of collections by testing for `IReadOnlyList<object>`. This does handle more types than before (previously only T[] and List<T> were handled) but I do not think the extension is problematic.
This also avoids the theoretical issue of the enumerator not being disposed if an exception is thrown during iteration, since we are using the indexer directly.
